### ISSUE TITLE
Rolling Cells "sometimes" causes problems with rep and route_emitter starting

### ISSUE THREAD

**Issue by andrew-edgar** [cloudfoundry/bpm-release#123]
We have seen problems when rolling a cell where the rep and route_emitter do not restart.

It looks like BPM issues as there are no logs in the rep logs but there are logs in the BPM logs.

I will attach some files with the details of the strace and also the BPM logs etc.

This is bpm 1.0.3
Diego 2.25.0

**Comment by andrew-edgar** [comment]
[debug.txt](https://github.com/cloudfoundry-incubator/bpm-release/files/2962658/debug.txt)
This is the log for strace -f bpm start rep

**Comment by andrew-edgar** [comment]
[debug2.txt](https://github.com/cloudfoundry-incubator/bpm-release/files/2962661/debug2.txt)
This is the log for strace -f bpm start route_emitter

**Comment by andrew-edgar** [comment]
route emitter failing logs in bpm.log ...

```
{"timestamp":"2019-03-13T16:51:37.101347957Z","level":"info","source":"bpm","message":"bpm.start.starting","data":{"job":"route_emitter","process":"route_emitter","session":"1"}}
{"timestamp":"2019-03-13T16:51:37.129054990Z","level":"error","source":"bpm","message":"bpm.start.failed-getting-job","data":{"error":"exit status 1","job":"route_emitter","process":"route_emitter","session":"1"}}
{"timestamp":"2019-03-13T16:51:37.129123152Z","level":"info","source":"bpm","message":"bpm.start.complete","data":{"job":"route_emitter","process":"route_emitter","session":"1"}}
{"timestamp":"2019-03-13T16:52:47.637712216Z","level":"info","source":"bpm","message":"bpm.start.acquiring-lifecycle-lock.starting","data":{"job":"route_emitter","process":"route_emitter","session":"1.1"}}
{"timestamp":"2019-03-13T16:52:47.637865983Z","level":"info","source":"bpm","message":"bpm.start.acquiring-lifecycle-lock.complete","data":{"job":"route_emitter","process":"route_emitter","session":"1.1"}}
{"timestamp":"2019-03-13T16:52:47.637896716Z","level":"info","source":"bpm","message":"bpm.start.starting","data":{"job":"route_emitter","process":"route_emitter","session":"1"}}
{"timestamp":"2019-03-13T16:52:47.663973865Z","level":"error","source":"bpm","message":"bpm.start.failed-getting-job","data":{"error":"exit status 1","job":"route_emitter","process":"route_emitter","session":"1"}}
{"timestamp":"2019-03-13T16:52:47.664067069Z","level":"info","source":"bpm","message":"bpm.start.complete","data":{"job":"route_emitter","process":"route_emitter","session":"1"}}
{"timestamp":"2019-03-13T16:53:57.179458063Z","level":"info","source":"bpm","message":"bpm.start.acquiring-lifecycle-lock.starting","data":{"job":"route_emitter","process":"route_emitter","session":"1.1"}}
{"timestamp":"2019-03-13T16:53:57.179625471Z","level":"info","source":"bpm","message":"bpm.start.acquiring-lifecycle-lock.complete","data":{"job":"route_emitter","process":"route_emitter","session":"1.1"}}
{"timestamp":"2019-03-13T16:53:57.179656933Z","level":"info","source":"bpm","message":"bpm.start.starting","data":{"job":"route_emitter","process":"route_emitter","session":"1"}}
{"timestamp":"2019-03-13T16:53:57.210749250Z","level":"error","source":"bpm","message":"bpm.start.failed-getting-job","data":{"error":"exit status 1","job":"route_emitter","process":"route_emitter","session":"1"}}
{"timestamp":"2019-03-13T16:53:57.210821488Z","level":"info","source":"bpm","message":"bpm.start.complete","data":{"job":"route_emitter","process":"route_emitter","session":"1"}}
```

**Comment by andrew-edgar** [comment]
I found /var/vcap/data/bpm/runc/bpm-rep/status.json was empty (same for route_emitter)

When I removed that file then I got a different error where when starting the rep I got the error message 

==> /var/vcap/sys/log/rep/rep.stderr.log <==
container with id exists: bpm-rep

**Comment by andrew-edgar** [comment]
working with Jim and Christopher they found ways to help fix it.

Basically step 1: remove the "empty" state.json in the `/var/vcap/data/bpm/runc/bpm-rep` and `/var/vcap/data/bpm/runc/bpm-route_emitter`

Then run this ...

/var/vcap/packages/bpm/bin/runc --root /var/vcap/data/bpm/runc delete -f bpm-rep
and
/var/vcap/packages/bpm/bin/runc --root /var/vcap/data/bpm/runc delete -f bpm-route_emitter

So it looks like "somehow" the state.json gets emptied and this causes problems with the cleanup

The root cause it to find out how and why the state.json becomes an empty file

**Comment by andrew-edgar** [comment]
[strace.txt](https://github.com/cloudfoundry-incubator/bpm-release/files/2963676/strace.txt)

This is the strace when I attempted to do the "runc" delete

**Comment by andrew-edgar** [comment]
This is happening more and more.  In our test env we have 16 cells and the latest update 3 of the cells failed to start because of the empty state.json in bpm-rep and bpm-route_emitter.  We are going to need this addressed very soon as this is a big issue to be able to deploy consistently.  what can I provide in the way of more information?

**Comment by andrew-edgar** [comment]
Now one env where this failed I'm getting this after trying to recover ...

```
/var/vcap/jobs/rep/bin/set-rep-kernel-params: line 21: /proc/sys/net/netfilter/nf_conntrack_max: No such file or directory
/var/vcap/jobs/rep/bin/set-rep-kernel-params: line 21: /proc/sys/net/netfilter/nf_conntrack_max: No such file or directory
/var/vcap/jobs/rep/bin/set-rep-kernel-params: line 21: /proc/sys/net/netfilter/nf_conntrack_max: No such file or directory
/var/vcap/jobs/rep/bin/set-rep-kernel-params: line 21: /proc/sys/net/netfilter/nf_conntrack_max: No such file or directory
/var/vcap/jobs/rep/bin/set-rep-kernel-params: line 21: /proc/sys/net/netfilter/nf_conntrack_max: No such file or directory
/var/vcap/jobs/rep/bin/set-rep-kernel-params: line 21: /proc/sys/net/netfilter/nf_conntrack_max: No such file or directory
/var/vcap/jobs/rep/bin/set-rep-kernel-params: line 21: /proc/sys/net/netfilter/nf_conntrack_max: No such file or directory
/var/vcap/jobs/rep/bin/set-rep-kernel-params: line 21: /proc/sys/net/netfilter/nf_conntrack_max: No such file or directory
/var/vcap/jobs/rep/bin/set-rep-kernel-params: line 21: /proc/sys/net/netfilter/nf_conntrack_max: No such file or directory
```
so something has messed up this cell completely and looks like it needs a recreate

**Comment by andrew-edgar** [comment]
looking a little deeper.  the state.json in bpm-rep in the most recent failure the  modification time is this ...
```
-rw-r--r--  1 root root    0 Mar 18 16:23 state.json
```
```
{"timestamp":"2019-03-18T16:22:45.275441127Z","level":"info","source":"bpm","message":"bpm.start.releasing-lifecycle-lock.starting","data":{"job":"rep","process":"rep","session":"1.3"}}
{"timestamp":"2019-03-18T16:22:45.275491262Z","level":"info","source":"bpm","message":"bpm.start.releasing-lifecycle-lock.complete","data":{"job":"rep","process":"rep","session":"1.3"}}
{"timestamp":"2019-03-18T16:23:22.219434323Z","level":"info","source":"bpm","message":"bpm.start.acquiring-lifecycle-lock.starting","data":{"job":"rep","process":"rep","session":"1.1"}}
{"timestamp":"2019-03-18T16:23:22.219616833Z","level":"info","source":"bpm","message":"bpm.start.acquiring-lifecycle-lock.complete","data":{"job":"rep","process":"rep","session":"1.1"}}
```
The logs above show around that time.  Is there an issue in the code where the release/acquire could clear out the state.json?

**Comment by xoebus** [comment]
```
/var/vcap/jobs/rep/bin/set-rep-kernel-params: line 21: /proc/sys/net/netfilter/nf_conntrack_max: No such file or directory
```

This looks like the `sysfs` is becoming unmounted? I would be very surprised if BPM was affecting the broader mounts on the system. Is there anything else happening on the system at the same time? Extremely high CPU load or disk pressure?

> Is there an issue in the code where the release/acquire could clear out the state.json?

The [locking code](https://github.com/cloudfoundry-incubator/bpm-release/blob/57e7db5d1d4919f8d4605f24d657c863001edcc6/src/bpm/commands/root.go#L138-L176) doesn't touch RunC at all. It does surround each start/stop/run operation though and so will appear correlated with operations which will read/write that file.

**Comment by andrew-edgar** [comment]
what is happening is this is a cell that we are rolling.  basically a small update in like bosh-dns-aliases.  all that is happening is ALL the processes are starting

**Comment by xoebus** [comment]
Looking at the strace output a little more, the only time that `state.json` is mentioned is when it's checked to see if it exists.

```
163480 lstat("/var/vcap/data/bpm/runc/bpm-rep/state.json", 0xc000207148) = -1 ENOENT (No such file or directory)
163480 openat(AT_FDCWD, "/var/vcap/data/bpm/runc/bpm-rep/state.json", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
```

I also noticed that during the `bpm-pre-start` that the entire download cache is `chown`ed. Is it possible that this is taking a long time (5-10s) and the process start is timing out?

**Comment by xoebus** [comment]
Scratch that, seems to be getting through that section fine.

It may be some errors in talking to D-Bus? `cat debug.txt  | grep 163656 | less` reveals that's the last thing it tries to do before the process stops.

**Comment by xoebus** [comment]
Could you please run the strace again on one of the cells with the `-t` and `-yy` options?

**Comment by jfmyers9** [comment]
Hey @andrew-edgar,

We spent some time looking at this and we are having a tough time reproducing the scenario as it's described. 

Is there any chance you could provide us with the following:
- the stemcell version
- the contents of `/var/vcap/sys/log/rep`

Also, it would be helpful to understand a bit more about the lifecycle of the incident. Can you describe what the failure you are seeing is in detail and what the status of `/var/vcap/data/bpm/runc` is at various points during the failure? (i.e. is the `state.json` empty after the `bosh deploy` fails?)

Also there were some recent changes to `bpm` to handle systemd more correctly. Would you be willing to try an older version of `bpm` and see if it exhibits similar behavior? (`v1.0.0`).

Best,

@jfmyers9 && @xoebus

**Comment by andrew-edgar** [comment]
the stemcell version is a SoftLayer stemcell  250.17.1 

/var/vcap/sys/log/rep I can get this nexxt time it fails.

the lifecycle is ...

We are doing a deploy.  everything was running ok before the deploy.
The cell is drained and stopped.
Once the source is updated monit starts everything up and rep and route_emitter cannot start.
At this point we see the state.json is empty.

Not sure when it becomes empty :) . It also from the bpm.log looks like during the time the cell is being drained and stopped.

Could this happen if the cell does not drain well?   if for some reason the drain timeout is exceeded and things need to be killed?

We would be willing to try any version you recommend :) . This does seem to have occurred since we updated to 1.0.3 from 1.0.0.

Andrew

**Comment by andrew-edgar** [comment]
we are going to revert back to 1.0.0.   we have also seen some weird issues where our cells have "crashed" after an update.  basically doing a bosh deployment the cell is updated fine and bosh moves on but then the cell "crashes" kernel panic or something.

We have only just started seeing this since we updated to 1.0.3 as well so we are reverting to investigate whether BPM is the cause of this panic as well.

**Comment by xoebus** [comment]
Thanks for letting us know. You may want to cherry pick the recent CVE commits into your release too.

We haven’t seen this issue on any other infrastructures. It may be worth bisecting through the recent commits to try and narrow down where the incompatibility is.

**Comment by jfmyers9** [comment]
@andrew-edgar We took a stab at fixing this after seeing it reported in another environment. We hope to release it in a `v1.0.4` as soon as it runs through our pipelines. Let us know if this release fixes the issue for you. You can find the fix at this commit here ab9293b93ec1fa16404b482c852f9163e56f5a3e.

Best,

@jfmyers9 && @xoebus
